src/net.c

```diff
@@ -999,6 +1003,110 @@ coap_send(coap_session_t *session, coap_pdu_t *pdu) {
   uint8_t r;
   ssize_t bytes_written;
 
+  if (pdu->code == COAP_RESPONSE_CODE(508)) {
+    /*
+     * Need to prepend our IP identifier to the data as per
+     * https://www.rfc-editor.org/rfc/rfc8768.html#section-4
+     */
+    char addr_str[INET6_ADDRSTRLEN + 8];
+    coap_opt_iterator_t opt_iter;
+    coap_opt_t *opt;
+    size_t hop_limit;
+
+    if (coap_print_addr(&session->addr_info.local, (uint8_t*)addr_str,
+                            INET6_ADDRSTRLEN + 8)) {
+      char *cp;
+      int len;
+
+      if (addr_str[0] == '[') {
+        cp = strchr(addr_str, ']');
+        if (cp) *cp = '\000';
+        if (memcmp(&cp[1], "::ffff:", 7)) {
+          /* IPv4 embedded into IPv6 */
+          cp = &addr_str[8];
+        }
+        else {
+          cp = &addr_str[1];
+        }
+      }
+      else {
+        cp = strchr(addr_str, ':');
+        if (cp) *cp = '\000';
+        cp = addr_str;
+      }
+      len = strlen(cp);
+
+      /* See if Hop Limit option is being used in return path */
+      opt = coap_check_option(pdu, COAP_OPTION_HOP_LIMIT, &opt_iter);
+      if (opt) {
+        coap_option_t result;
+
+        hop_limit =
+          coap_decode_var_bytes (coap_opt_value (opt), coap_opt_length (opt));
+        if (hop_limit == 1) {
+          coap_log(LOG_WARNING, "Proxy loop detected '%s'\n",
+                   (char*)pdu->data);
+          coap_delete_pdu(pdu);
+          return (coap_tid_t)COAP_DROPPED_RESPONSE;
+        }
+        else if (hop_limit < 1 || hop_limit > 255) {
+          /* Something is bad - need to drop this pdu (TODO or delete option) */
+          coap_log(LOG_WARNING, "Proxy return has bad hop limit count '%zd'\n",
+                                 hop_limit);
+          coap_delete_pdu(pdu);
+          return (coap_tid_t)COAP_DROPPED_RESPONSE;
+        }
+        hop_limit--;
```

Comment:
All the hop limit code should be handled by libcoap.  However, a proxy at an administrative domain border can elect to change / set up a hop limit value although this is not recommended.

Assuming the proxy passing back the 5.08 to the next downstream client(proxy) in the chain copies the HOP LIMIT option "as is", then the hop limit needs to be decremented by libcoap either on the incoming PDU from the upstream server(proxy), or as it is transmitted.  I went for the latter.  The addition of the IP address always has to be done on the latter so that the IP used to connect back to the downstream client is reported.

I can move the decrement to being handled on the incoming PDU if that makes better sense.

Reply:
Ok, thanks for clarifying. To me, decrementing on the outbound leg also seems more natural.